Introduce PostListViewModel and use PagedList from FluxC

oguzkocer opened the pull request:
**Note:** ~~This PR depends on https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/977 and should NOT be merged before that's merged and the FluxC hash is updated.~~

---

This PR refactors the post list by introducing `PostListViewModel` and moving all the logic out of the fragment and adapter to the view model. It aims to preserve the current design and behavior and achieves it except for a few minor details.

I have checked in with @malinajirka when I started working on this PR and we agreed to keep everything in a single PR. So, please don't yell at me `wpmobilebot`.

---

Generally speaking, I am not a big fan of refactoring existing components if the design or the requirements haven't changed for it. This was an exception for me for a few reasons:

1. When we implement filters and search, the logic will get even more complicated and that's a feature I'd like to see implemented sooner rather than later.
2. Post list is the only component currently using the `ListStore` and its implementation was for from exemplary. I didn't want people to get discouraged from using the `ListStore` just because of this.
3. Pagination can be a hard problem to solve and even after shipping the v1 of the post list that uses the `ListStore`, I didn't have enough confidence in it. It turns out I was right about, because as soon as I moved all the logic to the view model, I realized that calculating a whole list was very time consuming if the view was complicated enough and there are enough items. I've written about this a bit more in detail in the FluxC PR, so feel free to check it out here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/977

---

Most of the changes in this PR is following our current design patterns, so here are notes from other parts of it to be aware of:

* We are using `PagedListAdapter` from the [paging library](https://developer.android.com/topic/libraries/architecture/paging/) for the `PostListAdapter` since `ListStore` now exposes a `PagedList`. This moves the responsibility of how the changes are calculated and applied to the `PagedListAdapter` and other paging library components.
* As part of the changes to the FluxC for how the lists are consumed, we are transforming the `PostModel` to `PostAdapterItem` to make sure the expensive calculations are handled as part of pagination and in a background thread. When these calculations are scattered around in the adapter, it may not look like much, but they do take a considerable amount of time.
* I have separated different parts of post list into its own classes, reusable ones when applicable.
* I have created `ConnectionStatusLiveData` which can be injected into the view models to be able to handle network issues for actions such as trashing post, publishing post etc.
* `ToastMessageHolder` and `DialogHolder` was introduced to make it easier to show toasts and dialogs from view models.

### Known Issues

There are some issues I found while working on this component that's not specific to post list or this PR, but I think it's worth noting them here, so we can discuss them and maybe come up with solutions. Even if we can't fix them, it's good to have them documented somewhere for context.

* When the list is loading for the very first time and the user scrolls too fast, the replacement animation can look kind of wrong. Instead of the nice 1 to 1 replacement we get in most cases, it looks like the previous item is removed and the new one is added. This normally happens `areItemsTheSame` returns `false` for a row. My understanding is that this is due to paging library using `null` values when it can't process things fast enough and the internal diff calculation returns false before it's ever passed to us. If we were to disable the placeholders for the paging library, this wouldn't happen, but then we lose a lot more functionality we get from it such as showing the correct list size in initial load. I am skeptical if this will be something we can fix, but I am still trying out ideas as I think of them.
* Because we are using diff calculations, if an item is inserted at the very top, it'll be hidden and the only indication will be the flash of the scroll bar. This is not really specific to post list as page list and any other page that uses diffs will have the exact same issue, because that's what Google decided it should behave, but I find it annoying especially with UNDO of a trashed post.

I was able to fix the second issue in the previous implementation by saving and restoring the recyclerview state, but with `PagedListAdapter`, I didn't find this to be working well. I have tried _many_ other solutions and they all somewhat worked, but they were not perfect and had their own issues. Ideally, Google will change the behavior of the diff application to show the inserted data at the top if the list is at the top position. Worst case scenario, we can scroll the page manually in some cases, but that's a very error prone approach I'd like to avoid for now. I am leaving this to the post filters project for now.

### Questions and Check-ins

There are a few things I am not 100% sure about, so I am making a note of them here.

* If the post list is given a target post, we'll try to scroll to it after each refresh of the list until we can successfully do so for the first time. My question is, should we scroll to the target post even if it's not in the first loaded page? I haven't spent much time on this, because I _think_ we might be able to load that specific part of the list, but I was unable to take the time to explore it. I'll open a new issue for this if we don't figure it out during the review.

* I have moved the logic for dismissing the pending draft notification to the view model, except for the part where we dismiss it `onDetach` in some cases. I imagine this was added to handle configuration changes which is not very necessary with view models, but I was unable to test this properly as I was unable to create a pending draft notification. It'd be good to verify this during review.

---

I am copying most of the testing instructions and testing steps from #8473 and adding a `NEW` label for possible changes. I am hoping this will help put things in context.

### Testing Instructions

* Test a self-hosted site as well as a .com one. You don't have to test everything on both, but at least verify that basics work on both type of sites. **NEW:** I think the only rotation issue left is for the snackbars.

* Don't forget to rotate the screen at times. There are some rotation issues from before, but this PR shouldn't add any new ones.

* Use the emulator's "Cellular" settings to try a slower connection. (or take a subway ride 🤷‍) You don't have to test everything on a slow connection, but after you test everything on a fast connection, you can try the basics for a different site or after logging out and logging in on a slower one. If you slow the connection after everything is loaded once, it'll still load pretty fast, so won't be as useful of a test.

### Testing Steps

Since this touches everything in post list anything that affects it should be tested. Here is a non-comprehensive list: (only because I can't think of all of it)

* Do a fresh installation of the app (or logout/login) and visit the "Blog Posts" and check that it loads the posts.

_A note on fresh installations: This will be only time the posts are loaded from scratch. After this, only the ids and modified dates of the post will be fetched, which means your subsequent visits or pull to refresh will be a lot faster and more bandwidth efficient (+90% difference on both of them in my tests)_ **NEW:** This is still mostly correct, but performance for the initial load should be much better now.

* Tap on a post and change the title or description and go back. The post should be updated and should say "Local changes".

* Tap on a post, go to "Post settings" and add a featured image, tap on update button and check that it's reflected in the post list after a brief wait. (network request)

_It looks like this doesn't work if the change only happens locally. Check #8472 for more details._

* Tap on the floating button and create a draft by filling in title and/or content and going back. Check that it shows up in the list.

* Disable your network, tap on the floating button and create a local draft by filling in the title and/or content. Check that it shows up in the list. **NEW:** Also test that network actions are handled properly for publish and trash actions.

* Re-enable your network and tap on publish on the remote draft you created 2 steps ago. Check that it's showing that it's uploading and it's refreshed after the upload succeeds.

* Repeat the above step for the local draft you created.

* If you have a site with more than 100 posts, use that and scroll as far as you can go. (or as much as your patience allows) Check that more posts are loaded as you go and load more progress bar is visible at times. You can also overwrite the `pageSize` in `PostListDescriptor` in `PostListFragment` if you don't have a site with as many posts. **NEW:** You can update the `pageSize` from the view model now.

* Change a post from another client. (Calypso, iOS etc) Pull to refresh the list and check that the post is updated in the post list.

* Trash a post in the list and check that it's removed and a snackbar is shown to undo. Check that it's deleted in another client.

* Trash a post in the list and tap on the undo bar to check that it's re-added to the list and it's not deleted from remote. (verify on remote after waiting a minute)

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt:
```diff
@@ -171,25 +90,161 @@ class PostListFragment : Fragment(),
+            val targetLocalPostId = activity?.intent?.getIntExtra(EXTRA_TARGET_POST_LOCAL_ID, -1)?.let {
+                if (it != -1) it else null
+            }
+            viewModel = ViewModelProviders.of(postListActivity, viewModelFactory)
```
I think we should pass in an instance of the fragment instead of the activity.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/viewmodel/helpers/ConnectionStatusLiveData.kt:
```diff
@@ -0,0 +1,31 @@
+
+class ConnectionStatus(val isConnected: Boolean)
+
+class ConnectionStatusLiveData(private val context: Context) : LiveData<ConnectionStatus>() {
```
Fixed in 524239c8e7be7b0133c27fde35b57986dd83ea2c.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/PagedListDiffItemCallback.kt:
```diff
@@ -0,0 +1,49 @@
+    val getRemoteItemId: (T) -> Long?,
+    val areItemsTheSame: (T, T) -> Boolean,
+    val areContentsTheSame: (T, T) -> Boolean
+
```
Fixed in 524239c8e7be7b0133c27fde35b57986dd83ea2c.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/viewmodel/helpers/DialogHolder.kt:
```diff
@@ -0,0 +1,26 @@
+    @StringRes val positiveButtonTextRes: Int,
+    @StringRes val negativeButtonTextRes: Int
+) {
+    fun show(context: Context, fragmentManager: FragmentManager?) {
```
Fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/8587/commits/2475907b243f64e73acc1cb11c686fc64c7ea935.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -153,7 +79,7 @@ class PostListAdapter(
             }
             VIEW_TYPE_POST -> {
                 val view = layoutInflater.inflate(R.layout.post_cardview, parent, false)
```
I have given this some thought after our conversation and I do see the value in having the layout name in the holder to make it easy to navigate. However, I don't like to pass in more information to the holder than necessary. So, I think I'll keep it like this in this PR, but I am open to exploring this in the site creation project and piggybacking to the post list once we reach a consensus.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/adapters/PostListAdapter.kt:
```diff
@@ -1,143 +1,69 @@
-    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
-        super.onAttachedToRecyclerView(recyclerView)
-        this.recyclerView = recyclerView
+        postViewHolderConfig = PostViewHolderConfig(
```
Fixed in 2050fb161afb4cc1b2c1a34d1a6d505aeb580c70.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostListFragment.kt:
```diff
@@ -235,775 +280,88 @@ class PostListFragment : Fragment(),
-        ) {
-            UploadUtils.publishPost(nonNullActivity, post, site, dispatcher)
+        if (resultCode == Activity.RESULT_OK) {
+            viewModel.handleEditPostResult(data)
```
I completely agree with the logic, but in this case, it turns out there is no great way to avoid it. Since `UploadUtils.handleEditPostResultSnackbars` requires the data, we could either bypass the `ViewModel` completely or have to refactor the `UploadUtils`. I don't like either option and since this issue is testing related and there are no plans to add tests right now for it, I think it's best to keep it as is and address it when we can. As we discussed on the call, this PR doesn't have to fix everything in post list, it just has to take a step in the right direction.

oguzkocer commented:
Thanks for testing everything in detail @malinajirka! I've replied to all your comments and review notes. I believe the crash you mentioned in the first point is the only blocker we have and should be fixed now. Unless you find any other blockers, let's make sure this is merged as soon as possible since the FluxC PR we merged includes breaking changes. Feel free to open separate issues and assign them to me if you find any non-blocking issues.

> 1. The list keeps crashing with the following exception. At first, I thought one of the post is missing an attribute and it causes this exception. However, I'm not able to reproduce it in 100% cases.

This issue seems to be related to a `MediaModel` not having a `url`. I don't think any of the lists I have has this issue, that's why I missed it. I've added a proper null check for it in b31e9304b2c34907e6b3c2946db36150794f1bfb and should work fine now. Could you please test and verify?

> 2. Even when everything works as expected there are a lot of error responses in the LogCat -> is this expected behavior?

Yes that's expected. We currently fetch each item individually and FluxC generates those logs. I haven't added any logs myself. I am hoping we can switch to batched requests at a later time and I have made a request for the API support for it. However, it's outside the scope of this PR.

> 3. Request loop - items are being requested "in an ?infinite? loop" -> even going to the MySite fragment and back to the PostList fragment doesn't help ( I think this is related to the fact that we are passing an instance of the parent activity to the ViewModelProvider instead of an instance of the PostList fragment.)

I have fixed the `ViewModelProvider` issue in 7ca936a04c0c3e7f828392d543aaf5009026582a. Hopefully that fixes it as I have no idea how this is happening otherwise. Could you test again and see if it's fixed? If not, do you mind opening a new issue for it, so we don't block this PR? You can assign the issue to me. Thanks!

> Maybe this is also related?

I don't think it is, but again, I don't really know what's going on there.

> 4. I'm not sure how this should work, but I'll mention it anyway

> Issue 1: some of the items are loading indefinitely and there is no way how to retry except of pulling to refresh (maybe we should introduce an error list item with retry??

We have an issue for this in FluxC repo: https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/938. This will have to be handled at a later time.

> issue 2: I pulled to refresh and scrolled down -> there were no more items loading and it seemed like the complete PostList is shown, but there were a lot of items missing. Going to My Site and back to Post List solved the issue.

I think the proper way to solve this issue is to add a `LoadMoreIndicator` item to the list and show either a progress bar or a button to load more depending on the current state. That's in my plans for future enhancements and I'll open an issue for it soon. However it's outside the scope of this PR and same behavior should be present in `develop`.

> 5.Prerequisite - do not have any local drafts

> Issue 1: When the upload completes the uploading item is removed and the real item is inserted -> it's not replaced -> I have to manually scroll to the top as the real item is inserted "above the top border of the screen"

This depends on the internet connection, but yes, this can happen and it's mentioned in the known issues. It's a problem with diffutils and not specific to this PR. Having said that I do want to fix it, just not in this PR.

> Issue 2: When an item is uploading and I manually pull to refresh I can see two instances of the post.

I haven't been able to reproduce this, if you are able to reproduce this issue somewhat reliably, do you mind opening a new issue with the testing steps so I can fix it separately? Thanks!
    
> 6. When I remove a post and UNDO the action the item is inserted "above the top border of the screen" -> I have to manually scroll to it

This issue is mentioned in the known issues section.

> 7. trash/delete an item -> item disappears || trash/delete a second item before the UNDO action disappears -> the first item reappears (but in fact only the first item was removed -> pull to refresh and the first item disappears and the second item reappears)

Did I mention that I hated the undo implementation? For now, I've blocked the trash action until the first snackbar is removed in d6d0448034ea8d8bba21bb2b8aa5d985b51e8df1. This is not ideal, so I have opened a new issue for it to handle it separately. #8592

malinajirka commented:
None of the remaining issues seems critical + the current FluxC develop is not compatible with WPAndroid.

Thanks!

malinajirka commented:
>1. This issue seems to be related to a MediaModel not having a url. I don't think any of the lists I have has this issue, that's why I missed it. I've added a proper null check for it in b31e930 and should work fine now. Could you please test and verify?

👍 Seems ok

>2. Yes that's expected. We currently fetch each item individually and FluxC generates those logs. I haven't added any logs myself. I am hoping we can switch to batched requests at a later time and I have made a request for the API support for it. However, it's outside the scope of this PR.

I was just confused with the error severity of the message, but I agree it's not tied to this PR.


>3. I have fixed the ViewModelProvider issue in 7ca936a. Hopefully that fixes it as I have no idea how this is happening otherwise. Could you test again and see if it's fixed? If not, do you mind opening a new issue for it, so we don't block this PR? You can assign the issue to me. Thanks!

Now going to My Site and back to the Post Lists cancels the requests, but the original(tons of requests are being dispatched) issue is still there.

>4.Issue 1:  We have an issue for this in FluxC repo: wordpress-mobile/WordPress-FluxC-Android#938. This will have to be handled at a later time.

Perfect, thanks!

>4. Issue 2: I think the proper way to solve this issue is to add a LoadMoreIndicator item to the list and show either a progress bar or a button to load more depending on the current state. That's in my plans for future enhancements and I'll open an issue for it soon. However it's outside the scope of this PR and same behavior should be present in develop.

I'm a bit confused here. I thought it's not a valid state when the list thinks there are no more items even though there are, isn't it?

>5. Issue 1: This depends on the internet connection, but yes, this can happen and it's mentioned in the known issues. It's a problem with diffutils and not specific to this PR. Having said that I do want to fix it, just not in this PR.

👍 


>5. Issue 2: I haven't been able to reproduce this, if you are able to reproduce this issue somewhat reliably, do you mind opening a new issue with the testing steps so I can fix it separately? Thanks!

I created a new issue #8594. I'm able to reproduce it in 100% cases. Let me know if the steps to reproduce don;'t work for you. Thanks

>6. This issue is mentioned in the known issues section.

You are right, sorry. 

>7. Did I mention that I hated the undo implementation? For now, I've blocked the trash action until the first snackbar is removed in d6d0448. This is not ideal, so I have opened a new issue for it to handle it separately. #8592

👍

oguzkocer commented:
> Now going to My Site and back to the Post Lists cancels the requests, but the original(tons of requests are being dispatched) issue is still there.

Several dispatches happening at the same time is the expected behavior, at least for now. That shouldn't translate to actual requests, because I think we handle at most 10 requests at a time, but we do dispatch and queue things. We'll definitely improve on this, but the improvements we'll make depends on whether we'll have batch request support or not. If we won't we might end up queuing things ourselves.

It's important to keep in mind that most of the issues we talked about should mostly happen the first time a list is loaded. After that, there should only be a few missing items and making those requests together is not really that big of a deal. That's why I intentionally didn't focus on this part of the solution. Hope that makes sense!

malinajirka commented:
>It's important to keep in mind that most of the issues we talked about should mostly happen the first time a list is loaded. After that, there should only be a few missing items and making those requests together is not really that big of a deal. That's why I intentionally didn't focus on this part of the solution. Hope that makes sense!

👍 Thanks!
